libfoundation/src/foundation-value.cpp

```diff
@@ -1095,17 +1163,17 @@ void __MCValueFinalize(void)
     for(uindex_t i = 0; i < kMCValuePoolCount; i++)
         while(s_value_pools[i] . count > 0)
         {
-            __MCValue *t_value;
+            __MCFreedValue *t_value;
             t_value = s_value_pools[i] . values;
 
 #ifdef HAVE_VALGRIND
 			/* Valgrind support */
 			/* The first few bytes of the buffer actually contain the
 			 * address of the following buffer. */
-			VALGRIND_MAKE_MEM_DEFINED(t_value, sizeof (__MCValue *));
+			VALGRIND_MAKE_MEM_DEFINED(t_value, sizeof (__MCFreedValue));
 #endif /* HAVE_VALGRIND */
 
-			s_value_pools[i] . values = *(__MCValue **)t_value;
+			s_value_pools[i] . values = t_value;
```

Comment:
This should be:

    s_value_pools[i].values = t_value->next;

There's also a larger problem with `__MCValueFinalize()`. At the moment, the pools are destroyed, and then the unique values, false, true and null constants are destroyed.  However, those are all allocated *from the pools that were just destroyed*, leading to double-frees.  The fix is to move the pool deletion so that it's the last thing that occurs in `__MCValueFinalize()`